[docs] Add documentation for underscored attributes. (#37854)
swiftlang/swift: docs/ReferenceGuides/UnderscoredAttributes.md

# Underscored Attributes Reference

**WARNING:** This information is provided primarily for compiler and standard
library developers. Usage of these attributes outside of the Swift monorepo
is strongly discouraged.

The Swift reference has a chapter discussing [stable attributes][Attributes].
This document is intended to serve as a counter-part describing underscored
attributes, whose semantics are subject to change and most likely need to
go through the Swift evolution process before being stabilized.

The attributes are organized in alphabetical order.

> 

Author: Currently missing: `@_alwaysEmitIntoClient`, `@_disfavoredOverload`, `@_specializeExtension`, `@_marker`, `@_unsafeSendable, `@_unsafeMainActor`, `@_implicitSelfCapture`, `@_inheritActorContext`, `@_distributedActorIndependent`, `@_objc_non_lazy_realization`, `@_show_in_interface`, `@_staticInitializeObjCMetadata`.

## `@_alignment`

## `@_borrowed`

## `@_cdecl`

Similar to `@_silgen_name` but uses the C calling convention.
This attribute doesn't have very well-defined semantics.

For potential ideas on stabilization, see
[Formalizing `@cdecl`](https://forums.swift.org/t/formalizing-cdecl/40677).

> 

Reviewer: ## `@_disfavoredOverload`

Marks a declaration that the type checker should try to avoid using. When the
expression type checker is considering overloads, it will always prefer a solution
with fewer `@_disfavoredOverload` declarations over one with more of them.

This was designed to help with `ExpressibleBy*Literal` types. The type checker
strongly prefers to give literals their default type (e.g. `Int` for `ExpressibleByIntegerLiteral`,
`String` for `ExpressibleByStringLiteral`, etc.), so if you want your API to prefer some other
type but accept the default one too, you may need to mark the declaration taking the default
literal type with this attribute:

    extension LocalizedStringKey: ExpressibleByStringLiteral { ... }
    
    extension Text {
        // We want `Text("foo")` to use this initializer:
        init(_ key: LocalizedStringKey) { ... }
    
        // But without @_disfavoredOverload, it would use this one instead:
        @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }
    }

`@_disfavoredOverload` is a stopgap design and we want to eventually fix
the type checker behavior and then remove the attribute or make it a no-op.
Use it only to work around obviously broken overload resolution behavior,
not to arbitrarily manipulate overload priority, and be prepared for it to break
without warning.

## `@_dynamicReplacement(for:)`

Marks a function as the dynamic replacement for another `dynamic` function.
This is similar to method swizzling in other languages such as Objective-C,
except that the replacement happens at program start (or loading a shared
library), instead of at an arbitrary point in time.

For more details, see [Dynamic method replacement](https://forums.swift.org/t/dynamic-method-replacement/16619).

## `@_effects`

## `@_exported`

Use to mark an imported module to re-export all its declarations.
This attribute is most commonly used by overlays.

```
// module M
func f() {}

// module N
@_exported import M

// module P
import N
let () = N.f() // OK
```

## `@_fixed_layout`

## `@_hasInitialValue`

## `@_hasMissingDesignatedInitializers`

## `@_hasStorage`

## `@_implementationOnly`

Used to mark an imported module as an implementation detail.
This prevents types from that module being exposed in API
(types of public functions, constraints in public extension etc.)
and ABI (usage in `@inlinable` code).

## `@_implements`

## `@_inheritsConvenienceInitializers`

> ## `@_nonEphemeral`

Reviewer 2: ````
## `@_nonEphemeral`

Marks a function parameter that cannot accept a temporary pointer produced from
an inout-to-pointer, array-to-pointer, or string-to-pointer conversion. Such a
parameter may only accept a pointer that is guaranteed to outlive the duration
of the function call.

Attempting to pass a temporary pointer to an `@_nonEphemeral` parameter will
produce a warning:

```swift
func foo(@_nonEphemeral _ ptr: UnsafePointer<Int>) {}

func bar(_ ptr: UnsafePointer<Int>) {
  var x = 0

  // warning: inout expression creates a temporary pointer, but argument #1
  // should be a pointer that outlives the call to 'foo'
  foo(&x)

  // warning: passing '[Int]' to parameter, but argument #1 should be a pointer
  // that outlives the call to 'foo'
  foo([1, 2, 3])

  foo(ptr) // Okay
}
```

This attribute is primarily used within the standard library on the various
`UnsafePointer` initializers to warn users about the undefined behavior caused
by using a temporary pointer conversion as an argument:

```swift
func baz() {
  var x = 0

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr = UnsafePointer(&x)

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr2 = UnsafePointer([1, 2, 3])
}
```

This is because the temporary pointer conversion produces a pointer that is only
guaranteed to be valid for the duration of the call to the initializer, but will
be invalid once the initialized `UnsafePointer` is returned to the caller,
thereby leaving a dangling pointer.

Note we do guarantee that inout-to-pointer conversions on static and global
variables produce non-ephemeral pointers, as long as they are stored and have no
observers:

```swift
var global = 0

struct S {
  static var staticVar = 0
}

func baz() {
  let ptr = UnsafePointer(&global) // Okay
  let ptr2 = UnsafePointer(&S.staticVar) // Also okay
}
```

Additionally, if they are of a tuple or struct type, their stored members
without observers may also be passed inout as non-ephemeral pointers.

For more information, see [the educational note on temporary pointer usage](/userdocs/diagnostics/temporary-pointers.md).
````

from a module to one of the modules it imports without breaking clients.

## `@_private(sourceFile:)`

Fully bypasses access control, allowing access to private declarations
in the imported module. The imported module needs to be compiled with
`-Xfrontend -enable-private-imports` for this to work.

## `@_semantics`

Marks a function as having particular high-level semantics that are
specially recognized by the SIL optimizer.

## `@_silgen_name`

Changes the symbol name for a function, similar to an ASM label in C,
except that the platform symbol mangling (leading underscore on Darwin)
is maintained.

Since this has label-like behavior, it may not correspond to any declaration;
if so, it is assumed that the function is implemented in C.

A function defined by `@_silgen_name` is assumed to use the Swift ABI.

## `@_specialize(...)`

Forces generation of a specialized implementation for a generic declaration.

See [docs/Generics.rst][] for more details.

## `@_spi(spiName)`

Marks a declaration as SPI (System Programming Interface),
instead of API. Modules exposing SPI and using library evolution
generate an additional `.private.swiftinterface` file in addition

> to the usual `.swiftinterface` file.

Reviewer: IIRC, you need to actually pass a separate `-emit-private-module-interface-path`; otherwise it won't create the second file.

Clients can access SPI by marking the import as
`@_spi(spiName) import Module`. This makes it easy to find out which
clients are using certain SPIs by doing a textual search.

## `@_transparent`

Marks a function to be "macro-like", i.e., it is guaranteed to be inlined

> in debug builds.

Reviewer: Might want to link to https://github.com/apple/swift/blob/main/docs/TransparentAttr.md here.

# Underscored Attributes Reference

**WARNING:** This information is provided primarily for compiler and standard
library developers. Usage of these attributes outside of the Swift monorepo
is STRONGLY DISCOURAGED.

> The Swift reference has a chapter discussing [stable attributes][Attributes].

Reviewer 3: ```suggestion
The Swift reference has a chapter discussing
[stable attributes](https://docs.swift.org/swift-book/ReferenceManual/Attributes.html).
```

This document is intended to serve as a counter-part describing underscored
attributes, whose semantics are subject to change and most likely need to
go through the Swift evolution process before being stabilized.

The attributes are organized in alphabetical order.

## `@_alwaysEmitIntoClient`

Forces the body of a function to be emitted into client code.
Note that this is distinct from `@inline(__always)`;
it doesn't force inlining at call-sites,
it only means that the implementation is compiled into

> the module which use the code instead of the the module

Reviewer 3: ```suggestion
the module which uses the code instead of the the module
```

with the definition.

This means that `@_alwaysEmitIntoClient`
definitions are _not_ part of the defining module's ABI.

Most notably, default argument expressions are implicitly
`@_alwaysEmitIntoClient`, which means that adding
a default argument to a function which did not have one previously
does not break ABI.

> ## `@_alignment`

Reviewer 3: (Not in alphabetical order.)

## `@_borrowed`

## `@_cdecl`

Similar to `@_silgen_name` but uses the C calling convention.
This attribute doesn't have very well-defined semantics.

Type bridging is not done, so the parameter and return types should
correspond directly to types accessible in C.
In most cases, it is preferable to define a static method on an `@objc`
class instead of using `@_cdecl`.

For potential ideas on stabilization, see
[Formalizing `@cdecl`](https://forums.swift.org/t/formalizing-cdecl/40677).

> ## `@_distributedActorIndependent`

Reviewer 3: (Not in alphabetical order.)

## `@_disfavoredOverload`

Marks a declaration that the type checker should try to avoid using. When the
expression type checker is considering overloads, it will prefer a solution
with fewer `@_disfavoredOverload` declarations over one with more of them.

Historically, it was designed to help with `ExpressibleByXYZLiteral` types.
The type checker strongly prefers to give literals their default type
(e.g. `Int` for `ExpressibleByIntegerLiteral`,
`String` for `ExpressibleByStringLiteral`, etc.).
So if you want your API to prefer some other type but accept the default one
too, you may need to mark the declaration taking the default literal type with
this attribute:

    extension LocalizedStringKey: ExpressibleByStringLiteral { ... }

    extension Text {
        // We want `Text("foo")` to use this initializer:
        init(_ key: LocalizedStringKey) { ... }

        // But without @_disfavoredOverload, it would use this one instead,
        // because that lets it give the literal its default type:
        @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }

>     }

Reviewer 3: (Syntax highlighting?)

`@_disfavoredOverload` is a stop-gap design; eventually the type-checker
behavior should be fixed and the attribute should be removed/made into a no-op.

It's main use case is to work around clearly undesirable overload resolution,
not to arbitrarily manipulate overload priority.

## `@_dynamicReplacement(for:)`

Marks a function as the dynamic replacement for another `dynamic` function.
This is similar to method swizzling in other languages such as Objective-C,
except that the replacement happens at program start (or loading a shared
library), instead of at an arbitrary point in time.

For more details, see [Dynamic method replacement](https://forums.swift.org/t/dynamic-method-replacement/16619).

## `@_effects`

## `@_exported`

Use to mark an imported module to re-export all its declarations.
This attribute is most commonly used by overlays.

> ```
> // module M
> func f() {}
> 
> // module N
> @_exported import M
> 
> // module P
> import N
> let () = N.f() // OK

Reviewer 3: ```suggestion
```swift
// module M
func f() {}

// module N
@_exported import M

// module P
import N
let _ = N.f() // OK
```

```

> ## `@_fixed_layout`

Reviewer 3: https://forums.swift.org/t/swift-org-blog-library-evolution-in-swift/33785/14

Author: I have a fuzzy understanding, but I'd rather someone else with a clearer understanding describe it. It seems like this is used for both classes and protocols. cc @slavapestov ?

Author: Updated based on discussion with Slava.

## `@_hasInitialValue`

## `@_hasMissingDesignatedInitializers`

## `@_hasStorage`

## `@_implementationOnly`

Used to mark an imported module as an implementation detail.
This prevents types from that module being exposed in API
(types of public functions, constraints in public extension etc.)
and ABI (usage in `@inlinable` code).

## `@_implements`

## `@_implicitSelfCapture`

Allows access to `self` inside a closure without explicitly capturing it,
even when `Self` is a reference type.

> ```
> class C {
>   func f() {}
>   func g(_: @escaping () -> ()) {
>     g({ f() }) // error: call to method 'f' in closure requires explicit use of 'self'
>   }
>   func h(@_implicitSelfCapture _: @escaping () -> ()) {
>     h({ f() }) // ok
>   }
> }

Reviewer 3: ```suggestion
```swift
class C {
  func f() {}
  func g(_: @escaping () -> Void) {
    g({ f() }) // error: call to method 'f' in closure requires explicit use of 'self'
  }
  func h(@_implicitSelfCapture _: @escaping () -> Void) {
    h({ f() }) // ok
  }
}
```

```

## `@_inheritActorContext`

(Note that it is "inherit", not "inherits", unlike below.)

## `@_inheritsConvenienceInitializers`

## `@_marker`

Indicates that a protocol is a marker protocol;
one that represents some meaningful property at compile-time
but has no runtime representation.

> For more details, see [SE-0302](https://github.com/apple/swift-evolution/blob/main/proposals/0302-sendable-and-sendable-closures.md#marker-protocols), which introduces marker protocols.

Reviewer 3: (Pending apple/swift-evolution#1388)

```suggestion
For more details, see [SE-0302](https://github.com/apple/swift-evolution/blob/main/proposals/0302-concurrent-value-and-concurrent-closures.md#marker-protocols), which introduces marker protocols.
```

Author: Should we decide to merge that with changes, I'll merge it first. If we decide to not merge that change, I'll change this.

## `@_nonoverride`

## `@_objc_non_lazy_realization`

## `@_optimize(...)`

## `@_originallyDefinedIn`

Marks a declaration as being originally defined in a different module,
changing the name mangling. This can be used to move declarations
from a module to one of the modules it imports without breaking clients.

## `@_private(sourceFile:)`

Fully bypasses access control, allowing access to private declarations
in the imported module. The imported module needs to be compiled with
`-Xfrontend -enable-private-imports` for this to work.

## `@_semantics`

Marks a function as having particular high-level semantics that are
specially recognized by the SIL optimizer.

## `@_show_in_interface`

## `@_silgen_name`

Changes the symbol name for a function, similar to an ASM label in C,
except that the platform symbol mangling (leading underscore on Darwin)
is maintained.

Since this has label-like behavior, it may not correspond to any declaration;
if so, it is assumed that the function is implemented in C.

A function defined by `@_silgen_name` is assumed to use the Swift ABI.

## `@_specialize(...)`

Forces generation of a specialized implementation for a generic declaration.

> See [docs/Generics.rst][] for more details.

Reviewer 3: ```suggestion
See [Generics.rst](/docs/Generics.rst) for more details.
```

## `@_specializeExtension`

## `@_spi(spiName)`

Marks a declaration as SPI (System Programming Interface),
instead of API. Modules exposing SPI and using library evolution
generate an additional `.private.swiftinterface` file
(with `-emit-private-module-interface-path`) in addition
to the usual `.swiftinterface` file.

Clients can access SPI by marking the import as
`@_spi(spiName) import Module`. This makes it easy to find out which
clients are using certain SPIs by doing a textual search.

## `@_staticInitializeObjCMetadata`

## `@_transparent`

Marks a function to be "macro-like", i.e., it is guaranteed to be inlined
in debug builds.

> See [docs/TransparentAttr.md][] for more details.

Reviewer 3: ```suggestion
See [TransparentAttr.md](/docs/TransparentAttr.md) for more details.
```

# Underscored Attributes Reference

**WARNING:** This information is provided primarily for compiler and standard
library developers. Usage of these attributes outside of the Swift monorepo
is STRONGLY DISCOURAGED.

The Swift reference has a chapter discussing [stable attributes][Attributes].
This document is intended to serve as a counter-part describing underscored
attributes, whose semantics are subject to change and most likely need to
go through the Swift evolution process before being stabilized.

[Attributes]: https://docs.swift.org/swift-book/ReferenceManual/Attributes.html

The attributes are organized in alphabetical order.

> ## `@_alignment`

Reviewer 4: An attribute that allows the user to control the alignment of a type. The alignment value specified must be a power of two, and cannot be less than the "natural" alignment of the type that would otherwise be used by the Swift ABI. This attribute is intended for the SIMD types in the standard library which use it to increase their internal storage alignment to at least 16 bytes.

## `@_alwaysEmitIntoClient`

Forces the body of a function to be emitted into client code. Note that this is
distinct from `@inline(__always)`; it doesn't force inlining at call-sites, it
only means that the implementation is compiled into the module which uses the
code.

This means that `@_alwaysEmitIntoClient` definitions are _not_ part of the
defining module's ABI, so changing the implementation at a later stage
does not break ABI. (TODO: Does it matter if the `@_alwaysEmitIntoClient`

> definition is used in the defining module, say in an `@inlinable function?)

Author: I checked with a test file and it confirms what you said, it does creates a "non-external" symbol, not an external one.

gives the desired behavior:

```swift
extension LocalizedStringKey: ExpressibleByStringLiteral { ... }

extension Text {
  // We want `Text("foo")` to use this initializer:
  init(_ key: LocalizedStringKey) { ... }

  // But without @_disfavoredOverload, it would use this one instead,
  // because that lets it give the literal its default type:
  @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }
}
```

`@_disfavoredOverload` is a stop-gap design; eventually the type-checker
behavior should be fixed and the attribute should be removed/made into a no-op.

Its main use case is to work around clearly undesirable overload resolution,
not to arbitrarily manipulate overload priority.

## `@_dynamicReplacement(for:)`

Marks a function as the dynamic replacement for another `dynamic` function.
This is similar to method swizzling in other languages such as Objective-C,
except that the replacement happens at program start (or loading a shared
library), instead of at an arbitrary point in time.

For more details, see [Dynamic method replacement](https://forums.swift.org/t/dynamic-method-replacement/16619).

## `@_distributedActorIndependent`

> ## `@_effects`

Author: I've documented this for now. We can add more details later as needed.

## `@_exported`

Use to mark an imported module to re-export all its declarations.
This attribute is most commonly used by overlays.

```swift
// module M
func f() {}

// module N
@_exported import M

// module P
import N
func g() {
  N.f() // OK
}
```

## `@_fixed_layout`

## `@_hasInitialValue`

> ## `@_hasMissingDesignatedInitializers`

Reviewer 4: Proposing a definition here:

An attribute that indicates that there may be designated initializers that are not printed in the swiftinterface file for a particular class. This attribute is needed for the initializer model to maintain correctness when library evolution is enabled. This is because a class may have non-public designated initializers, and Swift allows the inheritance of convenience initializers if and only if the subclass overrides (or has synthesized overrides) of every designated initializer in its superclass. To see the problem in action, consider

```
// Lib.swift
open class A {
  init(invisible: ()) {}

  public init(visible: ()) {}
  public convenience init(hi: ()) { self.init(invisible: ()) }
}

// Client.swift
class B : A {
  var x: String
  
  public override init(visible: ()) {
    self.x = "Garbage"
    super.init(visible: ())
  }
}
```

In this case, if B were allowed to inherit the convenience initializer `A.init(invisible:)` then an instance created via `B(hi: ())` would fail to initialize `B.x` resulting in a memory safety hole. What's worse is there is no way to close this safety hole because the user cannot override the invisible designated initializer because they lack sufficient visibility.

## `@_hasStorage`

## `@_implementationOnly`

Used to mark an imported module as an implementation detail.
This prevents types from that module being exposed in API
(types of public functions, constraints in public extension etc.)
and ABI (usage in `@inlinable` code).

> ## `@_implements`

Reviewer 4: Proposed definition:

An attribute that indicates that a function with one name satisfies a protocol requirement with a different name. This is especially useful when two protocols declare a requirement with the same name, but the conforming type wishes to offer two separate implementations.

```
protocol P { func foo() }

protocol Q { func foo() }

struct S : P, Q {
  @_implements(P, foo())
  func foo_p() {}
  @_implements(Q, foo())
  func foo_q() {}
}
```

## `@_implicitSelfCapture`

Allows access to `self` inside a closure without explicitly capturing it,
even when `Self` is a reference type.

```swift
class C {
  func f() {}
  func g(_: @escaping () -> Void {
    g({ f() }) // error: call to method 'f' in closure requires explicit use of 'self'
  }
  func h(@_implicitSelfCapture _: @escaping () -> Void) {
    h({ f() }) // ok
  }
}
```

## `@_inheritActorContext`

(Note that it is "inherit", not "inherits", unlike below.)

> ## `@_inheritsConvenienceInitializers`

Reviewer 4: An attribute that signals that a class declaration inherits its convenience initializers from its superclass. More specifically, that all designated initializers - even those that may not be visible in a swiftinterface file - are overridden. This attribute is often printed alongside `@_hasMissingDesignatedInitializers` in this case.

```swift
func baz() {
  var x = 0

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr = UnsafePointer(&x)

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr2 = UnsafePointer([1, 2, 3])
}
```

The temporary pointer conversion produces a pointer that is only
guaranteed to be valid for the duration of the call to the initializer,
and becomes invalid once the call ends.
So the newly created `UnsafePointer` will be dangling.

One exception to this is that inout-to-pointer conversions
on static stored properties and global stored properties
produce non-ephemeral pointers, as long as they have no observers:

```swift
var global = 0

struct S {
  static var staticVar = 0
}

func baz() {
  let ptr = UnsafePointer(&global) // okay
  let ptr2 = UnsafePointer(&S.staticVar) // okay
}
```

Additionally, if they are of a tuple or struct type, their stored members
without observers may also be passed inout as non-ephemeral pointers.

For more details, see [the educational note on temporary pointer usage](/userdocs/diagnostics/temporary-pointers.md).

## `@_nonoverride`

## `@_objc_non_lazy_realization`

> ## `@_optimize(...)`

Reviewer 5: _optimize([none|size|speed]) controls the compiler's optimization mode. It is analagous to the command line options -Onone, -Osize, and -Ospeed, but limited to a single function body. `_optimizer(none)` can be used to diagnose compiler bugs or improve debugging in Release builds.

   type-check.
3. Move the necessary declarations from ToasterKit to ToasterKitCore.
   The moved declaration should have two attributes:
   - `@available` indicating when the declaration was introduced in ToasterKit.
   - `@_originallyDefinedIn` indicating the original module and when the
     declaration was moved to ToasterKitCore.
   ```
   @available(toasterOS 42, *)
   @_originallyDefinedIn(module: "ToasterKit", toasterOS 57)
   enum Toast {
   case underdone
   case perfect
   case burnt
   }
   ```
4. Add Swift compiler flags `-Xfrontend -emit-ldadd-cfile-path -Xfrontend /tmp/t.c`
   to ToasterKitCore's build settings. Add the emitted `/tmp/t.c` file to
   ToasterKit's compilation.
   This ensures when an app is built for deployment targets prior to the symbols' move,
   the app will look for these symbols in ToasterKit instead of ToasterKitCore.

## `@_private(sourceFile:)`

Fully bypasses access control, allowing access to private declarations
in the imported module. The imported module needs to be compiled with
`-Xfrontend -enable-private-imports` for this to work.

## `@_semantics`

Marks a function as having particular high-level semantics that are

> specially recognized by the SIL optimizer.

Reviewer 5: This is allows to optimizer to make use of some key invariants in performance critical data types, particularly Array. Since the implementation of those data types is written in Swift using Unsafe APIs, the optimizer would otherwise need to make conservative assumptions. Undefined behavior will result if the implementation changes in a way that violates the optimizer's assumptions about those semantics.

If an API should prefer some other type, but accept the default too,
marking the declaration taking the default type with `@_disfavoredOverload`
gives the desired behavior:

```swift
extension LocalizedStringKey: ExpressibleByStringLiteral { ... }

extension Text {
  // We want `Text("foo")` to use this initializer:
  init(_ key: LocalizedStringKey) { ... }

  // But without @_disfavoredOverload, it would use this one instead,
  // because that lets it give the literal its default type:
  @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }
}
```

`@_disfavoredOverload` is a stop-gap design; eventually the type-checker
behavior should be fixed and the attribute should be removed/made into a no-op.

Its main use case is to work around clearly undesirable overload resolution,
not to arbitrarily manipulate overload priority.

## `@_dynamicReplacement(for:)`

Marks a function as the dynamic replacement for another `dynamic` function.
This is similar to method swizzling in other languages such as Objective-C,
except that the replacement happens at program start (or loading a shared
library), instead of at an arbitrary point in time.

For more details, see

> [the forum post on Dynamic method replacement](https://forums.swift.org/t/dynamic-method-replacement/16619).

Author: This is deliberately longer for screen readers, since a screen reader jumping from link to link will only announce the link text.

> [the forum post on Dynamic method replacement](https://forums.swift.org/t/dynamic-method-replacement/16619).

Author: Unresolving, I will check with someone with more expertise on accessibility. It seems like articles on support.apple.com also use short link text in certain cases.

## `@_distributedActorIndependent`

## `@_effects(effectname)`

Tells the compiler that the implementation of the defined function is limited
to certain side effects. The attribute argument specifies the kind of side
effect limitations that apply to the function including any other functions
it calls. This is used to provide information to the optimizer that it can't
already infer from static analysis.

Changing the implementation in a way that violates the optimizer's assumptions
about the effects results in undefined behavior.

For more details, see [OptimizerEffects.rst](/docs/Proposals/OptimizerEffects.rst).

## `@_exported`

Re-exports all declarations from an imported module.

This attribute is most commonly used by overlays.

```swift
// module M
func f() {}

// module N
@_exported import M

// module P
import N
func g() {
  N.f() // OK
}
```

## `@_fixed_layout`

## `@_hasInitialValue`

## `@_hasMissingDesignatedInitializers`

Indicates that there may be designated initializers that are
not printed in the swiftinterface file for a particular class.

This attribute is needed for the initializer model to maintain correctness when

> library evolution is enabled. This is because a class may have non-public

Author: Could you describe why you think having a link specifically for library evolution would be valuable here? (There is one more place below mentioning library evolution under _spi, do you think I should link it there too?)

Generally, I'm in favor of having more links rather than fewer, but in this case, this document is not trying to explain much because it assumes that you mostly know what you're doing. E.g. it doesn't provide links for swiftinterfaces.

Also "Swift Library Evolution" gives pretty good search hits if you google it.

> library evolution is enabled. This is because a class may have non-public

Reviewer 6: We have a document about library evolution, if I recall, as a blog post; if a user is unfamiliar with the term it sounds similar to "Swift evolution" and can be be hard to google, in my view, and there's pretty low effort required to provide a link to the relevant document.

> library evolution is enabled. This is because a class may have non-public

Reviewer 6: I think our audience should strive to accommodate whoever might need to reference the meaning of the attribute after stumbling across it. So, `@_marker` is likely to have a broader audience than `@_staticInitializeObjCMetadata`.

> library evolution is enabled. This is because a class may have non-public

Author: Added link.

    self.x = "Garbage"
    super.init(visible: ())
  }
}
```

In this case, if `B` were allowed to inherit the convenience initializer
`A.init(invisible:)` then an instance created via `B(hi: ())` would fail
to initialize `B.x` resulting in a memory safety hole. What's worse is
there is no way to close this safety hole because the user cannot override
the invisible designated initializer because they lack sufficient visibility.

## `@_hasStorage`

## `@_implementationOnly`

Used to mark an imported module as an implementation detail.
This prevents types from that module being exposed in API
(types of public functions, constraints in public extension etc.)
and ABI (usage in `@inlinable` code).

## `@_implements(ProtocolName, Requirement)`

An attribute that indicates that a function with one name satisfies
a protocol requirement with a different name. This is especially useful
when two protocols declare a requirement with the same name, but the
conforming type wishes to offer two separate implementations.

```swift
protocol P { func foo() }

protocol Q { func foo() }

struct S : P, Q {
  @_implements(P, foo())
  func foo_p() {}
  @_implements(Q, foo())
  func foo_q() {}
}

> ```

Reviewer: It may not have come across, but I gave background on `@_disfavoredOverload` to try to illustrate when it should be used (i.e. to work around a known defect in overload resolution that can't be immediately fixed without breaking source, and particularly to work around that specific bug with literals).

argument:

```swift
func baz() {
  var x = 0

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr = UnsafePointer(&x)

  // warning: Initialization of 'UnsafePointer<Int>' results in a dangling pointer
  let ptr2 = UnsafePointer([1, 2, 3])
}
```

The temporary pointer conversion produces a pointer that is only
guaranteed to be valid for the duration of the call to the initializer,
and becomes invalid once the call ends.
So the newly created `UnsafePointer` will be dangling.

One exception to this is that inout-to-pointer conversions
on static stored properties and global stored properties
produce non-ephemeral pointers, as long as they have no observers:

```swift
var global = 0

struct S {
  static var staticVar = 0
}

func baz() {
  let ptr = UnsafePointer(&global) // okay
  let ptr2 = UnsafePointer(&S.staticVar) // okay
}
```

Additionally, if they are of a tuple or struct type, their stored members
without observers may also be passed inout as non-ephemeral pointers.

For more details, see [the educational note on temporary pointer usage](/userdocs/diagnostics/temporary-pointers.md).

> ## `@_nonoverride`

Reviewer 6: As per #19034:

#### `@_nonoverride`

The standard library and overlays are compiled with the `-warn-implicit-overrides` flag, which warns when a protocol restates a requirement from another protocol that it refines without annotating the declaration with either `override` or `@_nonoverride`.

An `override` annotation causes the overriding declaration to be treated identically to the overridden declaration; a conforming type can only provide one implementation ("witness"). Restating a protocol requirement and then marking it as an `override` is generally only needed to help associated type inference, and many `override` annotations correlate closely with ABI FIXMEs.

Meanwhile, `@_nonoverride` is the "opposite" of `override`, allowing two protocol requirements to be treated independently; a conforming type can provide a distinct witness for each requirement (for example, by using `@_implements`). Use `@_nonoverride` when semantics differ between the two requirements. For example, `BidirectionalCollection.index(_:offsetBy:)` allows negative offsets, while `Collection.index(_:offsetBy:)` does not, and therefore the former is marked `@_nonoverride`.

The `@_nonoverride` annotation can also be specified on class members in addition to protocol members. Since it is the "opposite" of `override`, it can be used to suppress "near-miss" diagnostics for declarations that are similar to but not meant to override another declaration, and it can be used to intentionally break the override chain, creating an overload instead of an override.

   runtime via ToasterKit, so existing apps continue to run.
2. In ToasterKit, use `@_exported import ToasterKitCore`.
   This ensures existing source code that only imports ToasterKit continues to
   type-check.
3. Move the necessary declarations from ToasterKit to ToasterKitCore.
   The moved declaration should have two attributes:
   - `@available` indicating when the declaration was introduced in ToasterKit.
   - `@_originallyDefinedIn` indicating the original module and when the
     declaration was moved to ToasterKitCore.
   ```swift
   @available(toasterOS 42, *)
   @_originallyDefinedIn(module: "ToasterKit", toasterOS 57)
   enum Toast {
   case underdone
   case perfect
   case burnt
   }
   ```
4. Add Swift compiler flags `-Xfrontend -emit-ldadd-cfile-path -Xfrontend /tmp/t.c`
   to ToasterKitCore's build settings. Add the emitted `/tmp/t.c` file to
   ToasterKit's compilation.
   This ensures when an app is built for deployment targets prior to the symbols' move,
   the app will look for these symbols in ToasterKit instead of ToasterKitCore.

More generally, mutliple availabilities can be specified, like so:

```swift
@available(toasterOS 42, bowlOS 54, mugOS 54, *)

> @_originallyDefinedIn(module: "ToasterKit", toasterOS 57, bowlOS 69, mugOS 69)

Reviewer 6: Is this (and the example above) backwards? It seems to read as though `Toast` was originally defined as part of ToasterKit in toasterOS 57, then sunk into ToasterKitCore in toasterOS 42.

If the example is correct and my reading is incorrect, perhaps the text here could explain explicitly which OS versions belong in which annotation.

Author: No, the examples are correct. The `@available` indicates since when the symbol is available. The OSes in `@_originallyDefinedIn` indicate when the symbol was moved to the current module.

There is text before this explaining the OS usage:

```
   - `@available` indicating when the declaration was introduced in ToasterKit.
   - `@_originallyDefinedIn` indicating the original module and when the
     declaration was moved to ToasterKitCore.
```

enum Toast { ... }
```

## `@_private(sourceFile: "FileName.swift")`

Fully bypasses access control, allowing access to private declarations
in the imported module. The imported module needs to be compiled with

> `-Xfrontend -enable-private-imports` for this to work.

Reviewer 6: Inquiring minds would want to know, what was this designed for/where is it used?

## `@_semantics("uniquely.recognized.id")`

Allows the optimizer to make use of some key invariants in performance critical
data types, especially `Array`. Since the implementation of these data types
is written in Swift using unsafe APIs, without these attributes the optimizer
would need to make conservative assumptions.

Changing the implementation in a way that violates the optimizer's assumptions
about the semantics results in undefined behavior.

> ## `@_show_in_interface`

Reviewer 6: Per commit 345d05e:

In general, standard library protocols whose names begin with underscores ("underscored protocols") are hidden in generated interfaces. The `@_show_in_interface` attribute is applied to underscored protocols in the standard library that should not be hidden.

(Is this even in active use? And is the flag to skip underscored stdlib protocols still enabled by default?)

Author: It's being used for `public protocol _AppendKeypath {}`. The field `SkipUnderscoredStdlibProtocols` is set to `true` in certain places, such as in `PrintOptions::printInterface`.

## `@_silgen_name("cName")`

Changes the symbol name for a function, similar to an ASM label in C,
except that the platform symbol mangling (leading underscore on Darwin)
is maintained.

Since this has label-like behavior, it may not correspond to any declaration;
if so, it is assumed that the function is implemented in C.

> A function defined by `@_silgen_name` is assumed to use the Swift ABI.

Reviewer 6: I believe there's another document explaining the different uses of `@_silgen_name` in more detail. Perhaps it could be linked to here?

Author: Add link to Standard Library Programmer's Manual.

## `@_specialize(...)`

Forces generation of a specialized implementation for a generic declaration.

See [Generics.rst](/docs/Generics.rst) for more details.

## `@_specializeExtension`

## `@_spi(spiName)`

Marks a declaration as SPI (System Programming Interface), instead of API.
Modules exposing SPI and using library evolution generate an additional
`.private.swiftinterface` file (with `-emit-private-module-interface-path`)
in addition to the usual `.swiftinterface` file. This private interface exposes
both API and SPI.

Clients can access SPI by marking the import as `@_spi(spiName) import Module`.
This design makes it easy to find out which clients are using certain SPIs by
doing a textual search.

## `@_staticInitializeObjCMetadata`

## `@_transparent`

Marks a function to be "macro-like", i.e., it is guaranteed to be inlined
in debug builds.

See [TransparentAttr.md](/docs/TransparentAttr.md) for more details.

> ## `@_typeEraser`

Reviewer 6: Per #29775 and #29966:

#### `@_typeEraser(Proto)`

Marks a concrete nominal type (class, enum, or struct) as one that implements type erasure for a protocol `Proto`.

The type must itself conform to `Proto`, and it must be visible everywhere that `Proto` is visible. It must also implement an initializer of the form `init<T: Proto>(erasing: T)` that is also visible everywhere `Proto` is visible; the initializer may have different generic constraints so long as it can be called with a value of any type that conforms to `Proto`.

This feature was designed to be used for compiler-driven type erasure for dynamic replacement of functions with an opaque return type.

## `@_weakLinked`

## `@_unsafeMainActor`, `@_unsafeSendable`

> Marks a parameter's (function) type as `@MainActor` (`@Sendable) in Swift 6 and

Reviewer 6: ```suggestion
Marks a parameter's (function) type as `@MainActor` (`@Sendable`) in Swift 6 and
```

within Swift 5 code that has adopted concurrency, but non-`@MainActor`

> (non-`@Sendable) everywhere else.

Reviewer 6: ```suggestion
(non-`@Sendable`) everywhere else.
```

> See the [forum post on Concurrency in Swift 5 and 6](https://forums.swift.org/t/concurrency-in-swift-5-and-6/49337)

Reviewer 6: ```suggestion
See the forum post [Concurrency in Swift 5 and 6](https://forums.swift.org/t/concurrency-in-swift-5-and-6/49337)
```

Author: Same as above, longer link text to provide better context to screen readers.

expression type checker is considering overloads, it will prefer a solution
with fewer `@_disfavoredOverload` declarations over one with more of them.

Historically, it was designed to help with `ExpressibleByXYZLiteral` types.
The type checker strongly prefers to give literals their default type
(e.g. `Int` for `ExpressibleByIntegerLiteral`,
`String` for `ExpressibleByStringLiteral`, etc.).
If an API should prefer some other type, but accept the default too,
marking the declaration taking the default type with `@_disfavoredOverload`
gives the desired behavior:

```swift
extension LocalizedStringKey: ExpressibleByStringLiteral { ... }

extension Text {
  // We want `Text("foo")` to use this initializer:
  init(_ key: LocalizedStringKey) { ... }

  // But without @_disfavoredOverload, it would use this one instead,
  // because that lets it give the literal its default type:
  @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }
}
```

`@_disfavoredOverload` is a stop-gap design; eventually the type-checker
behavior should be fixed and the attribute should be removed/made into a no-op.

Its main use case is to work around clearly undesirable overload resolution,

> not to arbitrarily manipulate overload priority.

Reviewer: I'm going to suggest changing the wording here to emphasize its role as a usage guideline:

> Use `@_disfavoredOverload` to work around known bugs in the overload resolution rules that cannot be immediately fixed without a source break. Don't use it to adjust overload resolution rules that are otherwise sensible but happen to produce undesirable results for your particular API; it will likely be removed or made into a no-op eventually, and then you will be stuck with an overload set that cannot be made to function in the way you intend.
> 
> `@_disfavoredOverload` was first introduced to work around a bug in overload resolution with `ExpressibleByXYZLiteral` types. The type checker strongly prefers to give literals their default type (e.g. `Int` for `ExpressibleByIntegerLiteral`, `String` for `ExpressibleByStringLiteral`, etc.). If an API should prefer some other type, but accept the default too, marking the declaration taking the default type with `@_disfavoredOverload` gives the desired behavior:
> 
>```swift
> extension LocalizedStringKey: ExpressibleByStringLiteral { ... }
> 
> extension Text {
>   // We want `Text("foo")` to use this initializer:
>   init(_ key: LocalizedStringKey) { ... }
> 
>   // But without @_disfavoredOverload, it would use this one instead,
>   // because that lets it give the literal its default type:
>   @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }
> }
>```

`````
Use `@_disfavoredOverload` to work around known bugs in the overload
resolution rules that cannot be immediately fixed without a source break.
Don't use it to adjust overload resolution rules that are otherwise sensible
but happen to produce undesirable results for your particular API; it will
likely be removed or made into a no-op eventually, and then you will be
stuck with an overload set that cannot be made to function in the way
you intend.

`@_disfavoredOverload` was first introduced to work around a bug in overload
resolution with `ExpressibleByXYZLiteral` types. The type checker strongly
prefers to give literals their default type (e.g. `Int` for 
`ExpressibleByIntegerLiteral`, `String` for `ExpressibleByStringLiteral`,
etc.). If an API should prefer some other type, but accept the default too,
 marking the declaration taking the default type with `@_disfavoredOverload`
 gives the desired behavior:

```swift
extension LocalizedStringKey: ExpressibleByStringLiteral { ... }

extension Text {
  // We want `Text("foo")` to use this initializer:
  init(_ key: LocalizedStringKey) { ... }

  // But without @_disfavoredOverload, it would use this one instead,
  // because that lets it give the literal its default type:
  @_disfavoredOverload init<S: StringProtocol>(_ str: S) { ... }
}
```
`````

Additionally, if they are of a tuple or struct type, their stored members
without observers may also be passed inout as non-ephemeral pointers.

For more details, see the educational note on
[temporary pointer usage](/userdocs/diagnostics/temporary-pointers.md).

## `@_nonoverride`

> Warns when a protocol restates a requirement from another protocol that

Reviewer 6: How about, "Marks a declaration that is not an override of another, required when compiling with the `-warn-implicit-overrides` flag, which warns when a protocol restates a requirement from another protocol that it refines without annotating the declaration with either `override` or `@_nonoverride`."

Author: Updated wording.

## `@_objc_non_lazy_realization`

Marks a class as being non-lazily (i.e. eagerly) [realized](/docs/Lexicon.md#realization).

This is used for declarations which may be statically referenced and wouldn't
go through the normal lazy realization paths. For example, the empty array
class must be non-lazily realized, because empty arrays are statically
allocated. Otherwise, passing the empty array object to other code without

> triggering realization would load to a read from uninitialized memory.

Reviewer 7: The problem is not uninitialized memory, but rather that most ObjC runtime APIs assume/require that classes are realized, and passing an unrealized class to them results in a crash because of that. I'd say, "Otherwise, passing the empty array object to other code without triggering realization could allow for the unrealized empty array class to be passed to ObjC runtime APIs which only operate on realized classes, resulting in a crash."

## `@_show_in_interface`

Shows underscored protocols from the standard library in the generated interface.

> By default, SourceKit hides underscored protocols from the generated swiftinterface,

Author: Testing it out, SourceKit does hide underscored protocols (and more generally, any underscored declarations) from the swiftinterface it shows in the Xcode UI. SourceKit also hides underscored modules from code completion. I don't think there's anything specific about the standard library for this attribute.

> By default, SourceKit hides underscored protocols from the generated swiftinterface,

Author: Updated wording.

> ## `@_specializeExtension`

Author: cc @aschwaighofer, can you review the description in this commit for `@_specializeExtension`?

Reviewer 8: Looks good.
